### PR TITLE
FIX MultiProc deadlock

### PR DESCRIPTION
**Pull request by ccraddock**
Reduced the complexity of the MultiProc processing to reduce likelihood for deadlock by:
- removing semaphores and replacing with event, this allows for a timeout in the wait that safeguards against deadlock
- setting results directly from apply_async callback rather than two stage processing, where a signal is sent from the callback, that wakes the multiproc.run() thread, which then sets the results -- this is where i think the previous deadlock was happening

Found deadlock where events would be cleared before they were handled, creating scenarios where thread waits for a signal that will never come -- fixed by changing the location from before the wait to after

Added missing close thread pool

### CODE REVIEW

File: nipype/pipeline/plugins/multiproc.py [hunk 1]
```diff
@@ -178,44 +193,54 @@ def _report_crash(self, node, result=None):
...
+
     def _send_procs_to_workers(self, updatehash=False, graph=None):
         """ Sends jobs to workers when system resources are available.
             Check memory (gb) and cores usage before running jobs.
         """
         executing_now = []
 
         # Check to see if a job is available
-        jobids = np.flatnonzero((self.proc_pending == True) & \
+        currently_running_jobids = np.flatnonzero((self.proc_pending == True) & \
```
Comment (satra) [inline nipype/pipeline/plugins/multiproc.py]: these are not running yet.

**Comment by chrisfilo** [comment]
There are few more places where tests need to be reenabled:
https://github.com/nipy/nipype/search?utf8=%E2%9C%93&q=issues%2F1692

File: nipype/pipeline/plugins/multiproc.py (hunk 1 above)
Comment (ccraddock) [inline nipype/pipeline/plugins/multiproc.py]: according to this they should be, or am I confused?

https://github.com/nipy/nipype/blob/a746633d72d3e861a4b1a52038dd40a590792e24/nipype/pipeline/plugins/base.py#L205

Comment (satra) [inline nipype/pipeline/plugins/multiproc.py]: that should be revised.

the following line sets a flag when a job is submitted.

https://github.com/nipy/nipype/blob/a746633d72d3e861a4b1a52038dd40a590792e24/nipype/pipeline/plugins/base.py#L371

Comment (ccraddock) [inline nipype/pipeline/plugins/multiproc.py]: :-)

**Comment by satra** [comment]
@ccraddock - we have enabled pytest now in master. can you merge with master and resolve the `test_multiproc*` files

**Pull request closed without merging**